---
title: "Analyzing rhythmic data with compareRhythms"
author: Bharath Ananthasubramaniam
date: 21 Oct 2020
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analyzing circadian data with compareRhythms}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This package is designed to find features that change their rhythm parameters (*amplitude* and *phase*) between the control and experimental groups. In this vignette, we will walk through two examples that show the basic application of the `compareRhythms` package to microarray and RNA-seq data. Nevertheless, any rhythmic data can be analyzed using this package.

### Usage summary
The analysis is run using a single function `compareRhythms()`. To execute this function, the three necessary ingredients are the timeseries data, the experimental design and parameters to choose and tune the method. The output of the function is a *data.frame* with the names of the differentially rhythmic features, the category it is classified under and optionally the rhythm parameters of the feature in the two groups. The differential rhythmicity categories are **gain** of, **loss** of, **change** of, or  **same** rhythms (with respect to the reference/control group). 


### Time series data and experimental design
Only two inputs are mandatory to run the workflows in this package. 

1. A single (numeric) *matrix* combining both timeseries datasets. The rows of this matrix are the features and the columns are the different samples. The *rownames* of this data matrix provide the `id` list for the features.

2. A *data.frame* specifying the experimental design (details regarding each sample) in order to interpret the data matrix. 
    i) There must be one row in this data.frame describing each sample (column) in the data matrix. 
    ii) This data.frame must contain a numeric column named `time` specifying the time associated with the sample and a *factor* column named `group` specifying whether a sample belongs to the control or the experimental group. This `group` variable must have only two levels with one chosen as the reference with respect to which the results are presented. 
    iii) Optionally, a column named `batch` can be used to specify a categorical (factor) covariate representing an independent or confounding variable (Note: covariates cannot be included in all methods. More on that below). Currently, there is no capability to include continuous covariates in the analysis.

### Choice of method
The package currently offers a choice of 6 different methods. We describe below the tuning parameters available for each method.

There are three parameters that are common to all methods.

- `period` : This positive number is the period of the rhythms, whose amplitude and phase must be compared between the two timeseries datasets. Defaults to 24. 

- `amp_cutoff` : Only features with a peak-to-trough amplitude greater than this positive number in at least one group are included in the differential rhythmicity results. 

- `just_classify` : This boolean flag specifies if the amplitude and phase estimates of each differential rhythmic feature in the two groups (`just_classify = FALSE`) must be returned in addition to the list of `id`s and their classification into the differential rhythmicity categories.

The different methods are:

1. **Model selection** (`method = "mod_sel"` (default)): The different categories of differentially rhythmic features are represented by linear regression models that are fit to the data. The best model/category is selected by an information theoretic `criterion`. The quality of fit of the best category to the data is fine tuned using `schwarz_wt_cutoff`.
    + `criterion` is used to select the desired information criterion to pick the best model/category. "bic" (default) selects Bayesian Information Criterion (BIC) and "aic" selects Akaike Information Criterion (AIC). BIC penalizes model size more than AIC and hence favors smaller models.
    + `schwarz_wt_cutoff` (default = 0.6) is a probability threshold for the weight of the best model/category. The weight of the best model/category (called Schwarz weight with BIC and Akaike weight with AIC) is the probability that the chosen category is the best category given the data and the other models/categories. Higher this number (between 0 and 1), more certain is the classification. But models that dont reach this threshold are left unclassified.
    
    <u>Use cases</u>: This method can be used on any normalized data. Technically, the assumption that the noise/errors is/are normal at each sample and independent needs to be acceptable.
    
2. **DODR** (`method = "dodr"`): Rhythmic features in either group are first identified using [rain](https://doi.org/doi:10.18129/B9.bioc.rain) followed by filtering by rhythm amplitude (`amp_cutoff`). The resulting subset of features are processed using [DODR](https://cran.r-project.org/package=DODR), an R package to find the differentially rhythmic features. The features are classified after the fact into the 4 categories listed above.
    + `rhythm_fdr` is the threshold for pre-filtering rhythmic features in either group based on the multiple testing corrected p-value from rain.
    + `compare_fdr` is the threshold for selecting differentially rhythmic features using the multiple testing corrected p-values from DODR.
    
    <u>Use cases</u>: This method can be used on any normalized data.

3. **limma** (`method = "limma"`): This is an implementation of **DODR** using the linear modeling approach for microarrays in [limma](https://doi.org/doi:10.18129/B9.bioc.limma). The analysis follows the same steps with the difference that the filtering for rhythmic features in either group can be accomplished in a single test and differential rhythmicity test is also achieved in the same framework by comparing rhythm parameters (amplitude and phase).
    + `rhythm_fdr` is the threshold for pre-filtering rhythmic features in either group based on the multiple testing corrected p-value.
    + `compare_fdr` is the threshold for selecting differentially rhythmic features using the multiple testing corrected p-values
    + `robust` is a boolean to make the noise estimates for individual features in limma robust against outliers (see `eBayes` in [limma](https://doi.org/doi:10.18129/B9.bioc.limma) for details).
    
    <u>Use cases</u>: This method is to be used only on log normalized microarray data (see limma for details).

4. **voom** (`method = "voom"`): This is a variation of **limma** to also process RNA-seq data with the same pipeline. Count data are preprocessed using `voom` and then analyzed using the **limma** method above.
    + `rhythm_fdr` (see **limma**)
    + `compare_fdr` (see **limma**)
    + `robust` (see **limma**)
    + `outliers` is a boolean to downweight outlier *samples* in the analysis. (see `voomWithQualityWeights` in [limma](https://doi.org/doi:10.18129/B9.bioc.limma) for details)
    
    <u>Use cases</u>: This method is to be used with count data from an RNA-seq experiment. Count data from aligment (STAR, TopHat2) followed by quantification (htseq-count, summarizeOverlaps, featureCounts) can be directly used. If [tximport](https://doi.org/doi:10.18129/B9.bioc.tximport) is used to import data, then use counts after setting `countsFromAbundance = "lengthScaledTPM" or "scaledTPM"` in the `tximport()` call. 

5. **DESeq2** (`method = "deseq2"`): This is the **limma** workflow adapted to process RNA-seq data according to [DESeq2](https://doi.org/10.18129/B9.bioc.DESeq2).
    + `rhythm_fdr` (see **limma**)
    + `compare_fdr` (see **limma**)
    + `length` is an optional *matrix* (with the same size as data) containing the average transcript length of each gene in each sample.
    
    <u>Use cases</u>: This method is to be used with count data from an RNA-seq experiment. Count data from aligment (STAR, TopHat2) followed by quantification (htseq-count, summarizeOverlaps, featureCounts) can be directly used. If [tximport](https://doi.org/doi:10.18129/B9.bioc.tximport) is used to import data, then use `counts` and `length` obtained from the `tximport()` call with `countsFromAbundance = "no"`.

6. **edgeR** (`method = "edger"`): This is the **limma** workflow adapted to process RNA-seq data according to [edgeR](https://doi.org/10.18129/B9.bioc.edgeR).
    + `rhythm_fdr` (see **limma**)
    + `compare_fdr` (see **limma**)
    + `length` (see **DESeq2**)
    
    <u>Use cases</u>: This method is to be used with count data from an RNA-seq experiment. Count data from aligment (STAR, TopHat2) followed by quantification (htseq-count, summarizeOverlaps, featureCounts) can be directly used. If [tximport](https://doi.org/doi:10.18129/B9.bioc.tximport) is used to import data, then use `counts` and `length` obtained from the `tximport()` call with `countsFromAbundance = "no"`.
    
    
### Example 1
 

```{r setup, include=TRUE}
library(compareRhythms)
```


